omnigan/trainer.py

```diff
@@ -796,38 +818,131 @@ def get_d_loss(self, multi_domain_batch, verbose=0):
-                fake = self.G.decoders[decoder][target_domain](z, cond)
-                fake_d = self.D[decoder][target_domain](fake)
-                real_d = self.D[decoder][source_domain](x)
-                fake_loss = self.losses["D"](fake_d, False)
-                real_loss = self.losses["D"](real_d, True)
-                disc_loss[decoder][target_domain] += fake_loss
-                disc_loss[decoder][source_domain] += real_loss
-
+            if "a" in self.opts.tasks or "t" in self.opts.tasks:
+                if self.opts.gen.t.use_spade:
+                    task_tensors = self.G.decode_tasks(z)
+                    cond = get_conditioning_tensor(x, task_tensors)
+                for i, source_domain in enumerate(batch_domain):
+                    target_domain = self.translation_map[source_domain]
+                    decoder = "a" if i == 0 else "t"
+                    fake = self.G.decoders[decoder][target_domain](z, cond)
+                    fake_d = self.D[decoder][target_domain](fake)
+                    real_d = self.D[decoder][source_domain](x)
+                    fake_loss = self.losses["D"](fake_d, False)
+                    real_loss = self.losses["D"](real_d, True)
+                    disc_loss[decoder][target_domain] += fake_loss
+                    disc_loss[decoder][source_domain] += real_loss
+
+                    if verbose > 0:
+                        print(f"Batch {batch_domain} > {decoder}: {source_domain} to real ")
+                        print(f"Batch {batch_domain} > {decoder}: {target_domain} to fake ")
+            if "m" in self.opts.tasks:
                 if verbose > 0:
-                    print(f"Batch {batch_domain} > {decoder}: {source_domain} to real ")
-                    print(f"Batch {batch_domain} > {decoder}: {target_domain} to fake ")
+                    print("Now training the ADVENT discriminator!")
+
+                for param in self.D["m"]["maskAdvent"].parameters():
+                    param.requires_grad = True
+                z_decode = self.G.decoders["m"](z)
+                z_decode = z_decode[:, 0, :, :]
+                z_prime = 1 - z_decode
+                prob = torch.stack([z_decode, z_prime]).transpose(0, 1)
+                prob = prob.detach()    
+                if batch_domain == "r":
+                    disc_loss["m"]["maskAdvent"] += self.losses["D.m.advent"](
+                        None,
+                        prob.to(self.device),
+                        self.target_label,
+                        self.D["m"]["maskAdvent"]
+                    )
+                elif batch_domain == "s":
+                    disc_loss["m"]["maskAdvent"] += self.losses["D.m.advent"](
+                        None,
+                        prob.to(self.device),
+                        self.source_label,
+                        self.D["m"]["maskAdvent"]
+                    )
+                else:
+                    raise Exception("Wrong Domain Input!")
```

Comment:
Remove that, not needed with `zerotensor`

Reply:
Could you pls remind me what do I need to remove here?